dot/core/service.go

```diff
@@ -544,56 +539,63 @@ func (s *Service) createBlockResponse(msg *network.BlockRequestMessage) (*networ
 	responseData := []*types.BlockData{}
 
 	for _, hash := range subchain {
-		data, err := s.blockState.GetBlockData(hash)
-		if err != nil {
-			return nil, err
-		}
 
 		blockData := new(types.BlockData)
 		blockData.Hash = hash
 
-		// TODO: checks for the existence of the following fields should be implemented once #596 is addressed.
+		// set defaults
+		blockData.Header = optional.NewHeader(false, nil)
+		blockData.Body = optional.NewBody(false, nil)
+		blockData.Receipt = optional.NewBytes(false, nil)
+		blockData.MessageQueue = optional.NewBytes(false, nil)
+		blockData.Justification = optional.NewBytes(false, nil)
 
 		// header
-		if msg.RequestedData&1 == 1 {
-			blockData.Header = data.Header
-		} else {
-			blockData.Header = optional.NewHeader(false, nil)
+		if blockRequest.RequestedData&1 == 1 {
+			log.Debug("msg.RequestedData & 1, GetHeader")
+			retData, err := s.blockState.GetHeader(hash)
+			if err == nil && retData != nil {
+				blockData.Header = retData.AsOptional()
+			}
 		}
-
 		// body
-		if (msg.RequestedData&2)>>1 == 1 {
-			blockData.Body = data.Body
-		} else {
-			blockData.Body = optional.NewBody(false, nil)
+		if blockRequest.RequestedData&2>>1 == 1 {
```

Comment:
hmm, no brackets around the `blockRequest.RequestedData&2`? let's have consistency with all the statements, don't think it makes a logic difference but still nice